### ISSUE TITLE
Output continuous probabilities & Transfer learning

### ISSUE THREAD

**Issue by az7jh2** [simongraham/hovernet_inference#27]
@simongraham Hi. Thanks for helping with displaying the hover-net nuclei segmentation result.

We evaluated the hover-net performance on TCGA and our own dataset. The performance of nuclei segmentation and classification is good on TCGA dataset, as the hover-net training dataset includes some TCGA data. The nuclei segmentation is acceptable on our own dataset, but the classification of nuclei type is not satisfying. I guess the reason is because of image quality as well as color normalization.

So now I have two new issues.

1. Can I get a continuous probabilities rather than a nuclei type from hover-net? Then I can pick some cancer nuclei with high confidence for downstream analysis.

2. If I want to do transfer learning based on the hover-net inference, and I do not want to change the model of nuclei segmentation. I think maybe updating some parameters of the output layer of classification model only would be ok. Is it a way to do it? 

Thanks.

**Comment by jgamper** [comment]
> The nuclei segmentation is acceptable on our own dataset, but the classification of nuclei type is not satisfying.

@az7jh2 are you observing that the model over-predicts neoplastic cells?

**Comment by az7jh2** [comment]
@jgamper Yes, the major problem is much connective nuclei are misclassfied as neoplastic nuclei

**Comment by simongraham** [comment]
Hi @az7jh2 - apologies for the delay in getting back to you on this.

There can be a range of reasons why the classification may not be as you would expect. First of all, I would recommend looking to see what the pixel resolution of your images are. We trained the images at 40x objective magnification (around 0.275 microns/ pixel). If you were to process images scanned at 20x, then you may get unexpected results. It is hard to make a comment on whether you need stain normalisation without seeing a sample of the images that you are using. During the training process, we used colour augmentation and used image tiles extracted from WSIs from a variety of different sources. Therefore, the model should fairly robust to stain appearance. Like @jgamper alluded to, the model trained on PanNuke seems to also be biased towards neoplastic cells. 

The best thing to do - like you mentioned is to fine tune the model on your own dataset. This feature will need to be added. If it is something of interest, I can add this in the next couple of weeks. Please can you confirm whether this would be a useful feature and then it can be added.

**Comment by simongraham** [comment]
Hi @az7jh2 - apologies for the delay in getting back to you on this.

There can be a range of reasons why the classification may not be as you would expect. First of all, I would recommend looking to see what the pixel resolution of your images are. We trained the images at 40x objective magnification (around 0.275 microns/ pixel). If you were to process images scanned at 20x, then you may get unexpected results. It is hard to make a comment on whether you need stain normalisation without seeing a sample of the images that you are using. During the training process, we used colour augmentation and used image tiles extracted from WSIs from a variety of different sources. Therefore, the model should fairly robust to stain appearance. Like @jgamper alluded to, the model trained on PanNuke seems to also be biased towards neoplastic cells. 

The best thing to do, like you mentioned, is to fine tune the model on your own dataset. This feature will need to be added. If it is something of interest, I can add this in the next couple of weeks. Please can you confirm whether this would be a useful feature and then it can be added.

**Comment by az7jh2** [comment]
@simongraham Thanks for your reply.

We use hover-net at 40x WSI images, so I think the scaling is not the cause of performance degrading. We will conduct a comprehensive evaluation of hover-net result on different images with various image quality. 

I think the two features I previously mentioned are very useful and need to be added.

1. Supporting probalilities output along with the nuclei type. Then we can plot a whole ROC curve instead of only one point. And we can decide whether to use nuclei with high confidence only for downstream analysis.

2. Supporting fine tune based on user's images. An accurate prediction model is often preferred to be integrated in medical routines, though we can only focus on high confident predictions for safety.

**Comment by simongraham** [comment]
@az7jh2 - thanks for this. We will look into implementing the above suggestions within the next few weeks. It should be done within the month of August.

**Comment by az7jh2** [comment]
@simongraham Thanks for the efforts to integrate the two new features. I know how busy you are, and I appreciate you still intend to finish the update.

Acctually the 1st feature (Output continuous probabilities) is currently more important for our project. So if you only have limited time to work on the new feature, it's helpful to only release the update of 1st feature first.

Thanks!

**Comment by simongraham** [comment]
Hi @az7jh2 ,

Thanks for letting us know. We will work on that S a priority.

**Comment by pranathivemuri** [comment]
Hi @simongraham Thanks for your work here! We have tried to use hovernet on some of our histology images and have seen that some of the cells that are distinctly nuclei are not getting detected as any type of the 6 cells. Wondering if there was a probability threshold or some sort of normalization on the pipeline that removed them. 

also wanted to check if the branch major_refactor is under construction (if so, do you have a guesstimate of when it would be merged to master)and if it would include more arguments to tweak as per this PR. 

Thanks a lot!

**Comment by simongraham** [comment]
Hi @pranathivemuri 

Apologies for the late reply. We have now merged the major_refactor branch and can be used. Here, you can use MoNuSAC and PanNuke checkpoints now. Within the next week we will be releasing a PyTorch version of the code, so keep tuned.
